src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp

```diff
@@ -54,82 +60,59 @@ class npc_calvin_montague : public CreatureScript
 
         void EnterCombat(Unit* /*who*/) override { }
 
-        void DamageTaken(Unit* pDoneBy, uint32 &uiDamage) override
+        void DamageTaken(Unit* /*attacker*/, uint32 &damage) override
         {
-            if (uiDamage > me->GetHealth() || me->HealthBelowPctDamaged(15, uiDamage))
+            if (damage > me->GetHealth() || me->HealthBelowPctDamaged(15, damage))
             {
-                uiDamage = 0;
-
-                me->RestoreFaction();
-                me->SetImmuneToPC(true);
+                damage = 0;
                 me->CombatStop(true);
-
-                m_uiPhase = 1;
-
-                if (pDoneBy->GetTypeId() == TYPEID_PLAYER)
-                    m_uiPlayerGUID = pDoneBy->GetGUID();
+                EnterEvadeMode();
+                _events.ScheduleEvent(EVENT_EMOTE_RUDE, Seconds(3));
             }
         }
 
         void UpdateAI(uint32 diff) override
         {
-            if (m_uiPhase)
-            {
-                if (m_uiPhaseTimer <= diff)
-                    m_uiPhaseTimer = 7500;
-                else
-                {
-                    m_uiPhaseTimer -= diff;
-                    return;
-                }
+            _events.Update(diff);
 
-                switch (m_uiPhase)
+            if (uint32 eventId = _events.ExecuteEvent())
+            {
+                switch (eventId)
                 {
-                    case 1:
+                    case EVENT_EMOTE_RUDE:
+                        me->HandleEmoteCommand(EMOTE_ONESHOT_RUDE);
+                        _events.ScheduleEvent(EVENT_TALK, Seconds(2));
+                        break;
+                    case EVENT_TALK:
                         Talk(SAY_COMPLETE);
-                        ++m_uiPhase;
+                        _events.ScheduleEvent(EVENT_DRINK, Seconds(5));
                         break;
-                    case 2:
-                        if (Player* player = ObjectAccessor::GetPlayer(*me, m_uiPlayerGUID))
+                    case EVENT_DRINK:
+                        if (Player* player = ObjectAccessor::GetPlayer(*me, _playerGUID))
                             player->AreaExploredOrEventHappens(QUEST_590);
-
+                        _playerGUID.Clear();
                         DoCast(me, SPELL_DRINK, true);
-                        ++m_uiPhase;
+                        _events.ScheduleEvent(EVENT_SET_QUESTGIVER_FLAG, Seconds(12));
+                        break;
+                    case EVENT_SET_QUESTGIVER_FLAG:
+                        me->SetFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_QUESTGIVER);
+                        _events.ScheduleEvent(EVENT_STAND, Seconds(3));
                         break;
-                    case 3:
-                        EnterEvadeMode();
+                    case EVENT_STAND:
+                        me->SetStandState(UNIT_STAND_STATE_STAND);
                         break;
                 }
-
-                return;
             }
 
             if (!UpdateVictim())
```

Comment:
Should be on top, no? (before events update)

Reply:
Events should be executed when out of combat. Some other scripts have this before DoMeleeAttackIfReady(), so I left it there.

Reply:
You can control it using "events.IsInPhase" like: https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Outland/BlackTemple/boss_illidan.cpp#L779

Reply:
I'm not sure I understand, sorry :P

This isn't a boss, this is a simple quest script for a NPC. There's no phase, no complex events or anything else. It should just do something when out of combat, and do melee attacks in combat. Stop.

All other similar scripts use the same structure - why is this one any different?

Reply:
Normally ccrs complain alot because of it.
But i agree with you btw....
It is tested, can be merged already?

Reply:
Yep, tested and working. I only have the [video](https://www.youtube.com/watch?v=5CjYDkAUPy0&feature=youtu.be&t=173) for reference, and with those fixes the quest works exactly as seen there.

I'd say it can be merged :)